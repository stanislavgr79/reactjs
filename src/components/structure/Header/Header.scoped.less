.header {
    background-image: url('@resources/images/intro.jpg');
    background-size: 100% auto;
    margin-bottom: 10px;
    position: relative;
    .header_top {
        display: flex;
        justify-content: space-between;

        .site_name {
            padding-top: 10px;
            margin: unset;
        }

        .btn_addMovie {
            margin-top: 15px;
        }
    }

    .header_search {

        form {
            padding-bottom: 150px;
        }

        .search_title {
            font-size: 40px;
            color: #FFFFFF;
            font-weight: 100;
            padding-left: 12%;
            padding-top: 3%;
        }

        .search_wraper {
            padding-top: 30px;
            display: flex;
            flex-direction: row;
            justify-content: center;

            .input {
                border-radius: 5px;
                width: 60%;
                margin-right: 10px;
            }
        }
    }
}

