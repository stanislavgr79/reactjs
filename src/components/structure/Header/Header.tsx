import React, { useCallback, useEffect, useState } from 'react';
import { useDispatch } from 'react-redux';
import ModelMoviePopup from '../../container/ModelMoviePopup';

import SiteName from '../../atom/SiteName';
import Button from '../../atom/Button';
import Input from '../../atom/Input';

import { updateSearchValue } from '../../../redux/actions/search-actions';
import { updateShowPopup } from '../../../redux/actions/movies-actions';

import './Header.scoped.less';

export default function Header(): JSX.Element {
  const dispatch = useDispatch();
  const [searchValue, setSearchValue] = useState('');
  const [showPopup, setShowPopup] = useState(false);

  useEffect(() => {
    dispatch(updateShowPopup(showPopup));
  }, [dispatch, showPopup]);

  const togglePopup = useCallback(() => {
    setShowPopup(!showPopup);
  }, [showPopup]);

  const handleChange = useCallback((e: any) => {
    setSearchValue(e.currentTarget.value);
  }, []);

  const onClick = useCallback(() => {
    dispatch(updateSearchValue(searchValue));
  }, [dispatch, searchValue]);

  return (
    <>
      <div className="header">
        <section className="header_top container-md">
          <SiteName />
          <Button
            onClick={() => togglePopup()}
            buttonType="button"
            className="btn_addMovie"
            label="+ ADD MOVIE"
          />
          {showPopup && <ModelMoviePopup closePopup={togglePopup} role="add" />}
        </section>
        <section className="header_search container-md">
          <div className="search_title">FIND YOUR MOVIE</div>
          <form onSubmit={(event) => event.preventDefault()}>
            <div className="search_wraper">
              <Input
                name="search"
                placeholder="What do you want to watch?"
                autoComplete="off"
                value={searchValue}
                id="search"
                onChange={handleChange}
              />
              <Button buttonType="submit" onClick={onClick} className="btn_search" label="SEARCH" />
            </div>
          </form>
        </section>
      </div>
    </>
  );
}
