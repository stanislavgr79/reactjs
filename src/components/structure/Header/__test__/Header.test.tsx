import React from 'react';
import renderer from 'react-test-renderer';
import { createMemoryHistory } from 'history';
import { Provider } from 'react-redux';

import Header from '../Header';

import ReactRouter from 'react-router';

import configureStore from 'redux-mock-store';
import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import storeReal from '../../../../redux/redux-store';

describe('Header', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  const mockLocation = {
    pathname: '/movies',
    search: '',
    state: null,
    hash: '',
  };
  const history = createMemoryHistory();
  history.location = mockLocation;

  jest.mock('../../../atom/SiteName', () => {
    return {
      __esModule: true,
      default: () => {
        return <p className="site_name"></p>;
      },
    };
  });
  jest.spyOn(ReactRouter, 'useLocation').mockReturnValue(mockLocation);
  const replaceHistorySpy = jest.spyOn(ReactRouter, 'useHistory').mockReturnValue(history);
  const mockStore = configureStore();
  let store;

  test('snapshot renders', () => {
    store = mockStore();

    const tree = renderer.create(
      <Provider store={store}>
        <Header />
      </Provider>,
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });

  test('search input value display test1', () => {
    store = mockStore();

    render(
      <Provider store={store}>
        <Header />
      </Provider>,
    );

    fireEvent.change(screen.getByRole('textbox'), {
      target: { value: 'film' },
    });

    expect(screen.getByDisplayValue('film')).toBeInTheDocument();
  });

  test('search input value display test2', async () => {
    store = mockStore();

    render(
      <Provider store={store}>
        <Header />
      </Provider>,
    );

    expect(screen.queryByText(/film/)).toBeNull();

    await userEvent.type(screen.getByRole('textbox'), 'film');

    expect(screen.getByDisplayValue('film')).toBeInTheDocument();
  });

  test('handleChange callback', async () => {
    store = mockStore();

    const tree = render(
      <Provider store={store}>
        <Header />
      </Provider>,
    );

    fireEvent.change(screen.getByRole('textbox'), {
      target: { value: 'film' },
    });

    expect(tree.baseElement).toMatchSnapshot();
  });

  test('click search button', async () => {
    const store = storeReal;

    const tree = render(
      <Provider store={store}>
        <Header />
      </Provider>,
    );

    fireEvent.change(screen.getByRole('textbox'), {
      target: { value: 'film' },
    });
    fireEvent.click(screen.getByText('SEARCH'));

    expect(tree.baseElement).toMatchSnapshot();
    expect(replaceHistorySpy).toHaveBeenCalled();
    expect(store.getState().searchStore.search.value).toEqual('film');
  });

  test('click add button', async () => {
    store = mockStore();

    const tree = render(
      <Provider store={store}>
        <Header />
      </Provider>,
    );

    fireEvent.click(screen.getByText('+ ADD MOVIE'));

    expect(tree.baseElement).toMatchSnapshot();
    expect(tree.baseElement).toContainHTML('<div class="container add_movie">');
  });
});
