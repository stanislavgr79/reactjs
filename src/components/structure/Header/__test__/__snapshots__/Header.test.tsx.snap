// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header click add button 1`] = `
<body>
  <div>
    <div
      class="header"
    >
      <section
        class="header_top container-md"
      >
        <p
          class="site_name"
        >
          <strong>
            netflix
          </strong>
          roulette
        </p>
        <button
          class="btn button btn_addMovie"
          id=""
          type="button"
        >
          + ADD MOVIE
        </button>
        <div
          class="popup_model"
        >
          <div
            class="model_content container-md"
          >
            <p
              class="site_name"
            >
              <strong>
                netflix
              </strong>
              roulette
            </p>
            <div
              class="container add_movie"
            >
              <div
                class="container-sm form"
              >
                <div
                  class="page_close"
                >
                  <button
                    class="btn button btn_close"
                    id=""
                    type="button"
                  />
                </div>
                <div
                  class="page_title"
                >
                  ADD MOVIE
                </div>
                <form
                  class="movie_form"
                >
                  <label>
                    TITLE
                  </label>
                  <input
                    class="input simple_input"
                    name="title"
                    placeholder="Write Title"
                    type="text"
                    value=""
                  />
                  <label>
                    RELEASE DATE
                  </label>
                  <div
                    class="react-datepicker-wrapper"
                  >
                    <div
                      class="react-datepicker__input-container"
                    >
                      <input
                        class="simple_input"
                        name="release_date"
                        placeholder="Select Date"
                        readonly=""
                        value=""
                      />
                      <img
                        alt=""
                        class="image_calendar"
                        src="[object Object]"
                      />
                    </div>
                  </div>
                  <label>
                    POSTER URL
                  </label>
                  <input
                    class="input simple_input"
                    name="poster_path"
                    placeholder="Poster URL here"
                    type="text"
                    value=""
                  />
                  <label>
                    GENRE
                  </label>
                  <div
                    class="select_genres css-2b097c-container"
                  >
                    <div
                      class=" css-yk16xz-control"
                    >
                      <div
                        class=" css-g1d714-ValueContainer"
                      >
                        <div
                          class=" css-1wa3eu0-placeholder"
                        >
                          Select Genre
                        </div>
                        <div
                          class="css-b8ldur-Input"
                        >
                          <div
                            class=""
                            style="display: inline-block;"
                          >
                            <input
                              aria-autocomplete="list"
                              autocapitalize="none"
                              autocomplete="off"
                              autocorrect="off"
                              id="react-select-2-input"
                              spellcheck="false"
                              style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                              tabindex="0"
                              type="text"
                              value=""
                            />
                            <div
                              style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                            />
                          </div>
                        </div>
                      </div>
                      <div
                        class=" css-1hb7zxy-IndicatorsContainer"
                      >
                        <span
                          class=" css-1okebmr-indicatorSeparator"
                        />
                        <div
                          aria-hidden="true"
                          class=" css-tlfecz-indicatorContainer"
                        >
                          <svg
                            aria-hidden="true"
                            class="css-tj5bde-Svg"
                            focusable="false"
                            height="20"
                            viewBox="0 0 20 20"
                            width="20"
                          >
                            <path
                              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                    <div>
                      <input
                        name="genres"
                        type="hidden"
                      />
                    </div>
                  </div>
                  <label>
                    OVERVIEW
                  </label>
                  <input
                    class="input simple_input"
                    name="overview"
                    placeholder="Overview here"
                    type="text"
                    value=""
                  />
                  <label>
                    RUNTIME
                  </label>
                  <input
                    class="input simple_input"
                    min="0"
                    name="runtime"
                    placeholder="Runtime here"
                    type="number"
                    value=""
                  />
                  <div
                    class="button_section"
                  >
                    <button
                      class="btn button btn_submit"
                      id="add_movie_btn_submit"
                      type="submit"
                    >
                      SUBMIT
                    </button>
                    <button
                      class="btn button btn_reset"
                      id="add_movie_btn_reset"
                      type="reset"
                    >
                      RESET
                    </button>
                  </div>
                </form>
              </div>
            </div>
            <div
              class="footer"
            >
              <p
                class="site_name"
              >
                <strong>
                  netflix
                </strong>
                roulette
              </p>
            </div>
          </div>
        </div>
      </section>
      <section
        class="header_search container-md"
      >
        <div
          class="search_title"
        >
          FIND YOUR MOVIE
        </div>
        <form>
          <div
            class="search_wraper"
          >
            <input
              autocomplete="off"
              class="input"
              id="search"
              name="search"
              placeholder="What do you want to watch?"
              type="text"
              value=""
            />
            <button
              class="btn button btn_search"
              id=""
              type="submit"
            >
              SEARCH
            </button>
          </div>
        </form>
      </section>
    </div>
  </div>
</body>
`;

exports[`Header click search button 1`] = `
<body>
  <div>
    <div
      class="header"
    >
      <section
        class="header_top container-md"
      >
        <p
          class="site_name"
        >
          <strong>
            netflix
          </strong>
          roulette
        </p>
        <button
          class="btn button btn_addMovie"
          id=""
          type="button"
        >
          + ADD MOVIE
        </button>
      </section>
      <section
        class="header_search container-md"
      >
        <div
          class="search_title"
        >
          FIND YOUR MOVIE
        </div>
        <form>
          <div
            class="search_wraper"
          >
            <input
              autocomplete="off"
              class="input"
              id="search"
              name="search"
              placeholder="What do you want to watch?"
              type="text"
              value="film"
            />
            <button
              class="btn button btn_search"
              id=""
              type="submit"
            >
              SEARCH
            </button>
          </div>
        </form>
      </section>
    </div>
  </div>
</body>
`;

exports[`Header handleChange callback 1`] = `
<body>
  <div>
    <div
      class="header"
    >
      <section
        class="header_top container-md"
      >
        <p
          class="site_name"
        >
          <strong>
            netflix
          </strong>
          roulette
        </p>
        <button
          class="btn button btn_addMovie"
          id=""
          type="button"
        >
          + ADD MOVIE
        </button>
      </section>
      <section
        class="header_search container-md"
      >
        <div
          class="search_title"
        >
          FIND YOUR MOVIE
        </div>
        <form>
          <div
            class="search_wraper"
          >
            <input
              autocomplete="off"
              class="input"
              id="search"
              name="search"
              placeholder="What do you want to watch?"
              type="text"
              value="film"
            />
            <button
              class="btn button btn_search"
              id=""
              type="submit"
            >
              SEARCH
            </button>
          </div>
        </form>
      </section>
    </div>
  </div>
</body>
`;

exports[`Header snapshot renders 1`] = `
<div
  className="header"
>
  <section
    className="header_top container-md"
  >
    <p
      className="site_name"
    >
      <strong>
        netflix
      </strong>
      roulette
    </p>
    <button
      className="btn button btn_addMovie"
      id=""
      onClick={[Function]}
      type="button"
    >
      + ADD MOVIE
    </button>
  </section>
  <section
    className="header_search container-md"
  >
    <div
      className="search_title"
    >
      FIND YOUR MOVIE
    </div>
    <form
      onSubmit={[Function]}
    >
      <div
        className="search_wraper"
      >
        <input
          autoComplete="off"
          className="input"
          id="search"
          name="search"
          onChange={[Function]}
          placeholder="What do you want to watch?"
          type="text"
          value=""
        />
        <button
          className="btn button btn_search"
          id=""
          onClick={[Function]}
          type="submit"
        >
          SEARCH
        </button>
      </div>
    </form>
  </section>
</div>
`;
